library/core/src/ptr/const_ptr.rs

```diff
@@ -23,11 +23,12 @@ impl<T: ?Sized> *const T {
     /// assert!(!ptr.is_null());
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
+    #[rustc_const_unstable(feature = "const_ptr_is_null", issue = "74939")]
     #[inline]
-    pub fn is_null(self) -> bool {
+    pub const fn is_null(self) -> bool {
         // Compare via a cast to a thin pointer, so fat pointers are only
         // considering their "data" part for null-ness.
-        (self as *const u8) == null()
+        (self as *const u8).guaranteed_eq(null())
```

Comment:
My concern remains that this needs very good documentation to explain that during CTFE, the return value is in some sense "approximate".

Reply:
I added documentation, but this is indeed not too great. I mean, you'd have to do something very messy with a pointer in order to end up with a pointer that is a null pointer at runtime but not at compile-time. In fact, you can't offset a pointer beyond its allocation in const eval right now, no matter what kind of fancy unsafe code you employ (see the new test). So right now, this is not an issue.

Reply:
> In fact, you can't offset a pointer beyond its allocation in const eval right now

Certainly we want `wrapping_offset` to be `const fn` at some point.
But for the even nearer future, I think you can convert your pointer to `*const [u8; 4096]` and then to `ptr::raw_const!((*ptr)[1024])`, I think that should offset out-of-bounds.